src/net.cpp

```diff
@@ -2635,25 +2635,34 @@ void CConnman::StopNodes()
         }
     }
 
-    // Close sockets
-    LOCK(cs_vNodes);
-    for (CNode* pnode : vNodes)
-        pnode->CloseSocketDisconnect();
-    for (ListenSocket& hListenSocket : vhListenSocket)
-        if (hListenSocket.socket != INVALID_SOCKET)
-            if (!CloseSocket(hListenSocket.socket))
-                LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
+    // Delete peer connections
+    std::vector<CNode*> nodes;
+    {
+        LOCK(cs_vNodes);
+        nodes = std::move(vNodes);
+        vNodes.clear();
```

Comment:
Could write this as:

```c++
cs_vNodes.lock();
std::vector<CNode*> nodes(std::move(vNodes)); // move constructor clears vNodes
cs_vNodes.unlock();
``` 

If I'm reading the spec right, the move constructor is guaranteed to be constant time, while `operator=` is linear time; and the move constructor also guarantees the original ends up cleared. Since C++17 the move constructor is also marked noexcept, so `lock` and `unlock` in place of RAII locks should be sound. Alternatively, could use `WAIT_LOCK(cs_vNodes, lock); ...; REVERSE_LOCK(lock);`.

Reply:
Does performance matter here at all? Shutdown only happens once and flushing to disk will take magnitudes longer anyway.

Reply:
No I don't think so -- at worst it's just moving pointers around, and only a hundred or so of them in normal configurations.

Reply:
> the move constructor is guaranteed to be constant time, while operator= is linear time

This is the move assignment operator (number 2 in https://en.cppreference.com/w/cpp/container/vector/operator%3D). Its complexity is linear in the size of `this` - the vector being moved to. Presumably that's because the objects in `this` need to be destructed and the memory already owned by `this` needs to deallocated. In our case, `this` is empty, and so the operation is constant time - it just needs to copy start pointer/size/capacity from `other` to `this`.

> Does performance matter here at all?

Absolutely not.

> I wasn't sure if std::move + clear was safe / necessary

Moving from a vector leaves it in a "valid but unspecified state". The `clear()` is _probably_ unnecessary since this is shutdown and we're not going to touch `vNodes` again, but I think it's good practice to not to leave vectors in an unspecified state if we can help it.